Multicast Support - Work In Progress

evpopov opened the pull request:
I'm starting a PR in order to quit procrastinating on this subject and to align my ideas with the rest of the community. This is my proposal for how multicast can be added. Please feel free to comment.
This is nowhere near done, but it is functional and I am using it in my projects. 
If you want to to try this, you will have to modify your network driver to receive multicast packets. I only have ATSAME70 hardware so that is the only network driver that this PR updates.
As of the time of this PR's creation, the code only supports IPv4. It also supports multiple groups per socket, but this will be removed in the interest of lower CPU load. The goal is for 1 socket to only be able to subscribe to 1 multicast group.

Commit details:
Adds 2 function pointers to the network interface struct that handle adding and removing  multicast MAC addresses. Updates IGMP to use function pointers through the network interface. Makes the Add/Remove Multicast functions private to NetworkInterface.c They are now used through pointers in the NetworkInterface_t struct. Improves the SAME70 driver to handle adding/removing muticast MAC addresses

shubnil commented:
Thanks @evpopov for creating the PR. As discussed we will be actively reviewing the PR and keep you updated on the review status.

rawalexe commented:
Hello @evpopov ,
There are few conflicting files and a bit feed back from ci/cd related to build-checks, can you please apply these requested changes to your code at your convenience so that I can pass this to our team for review.

Thank you,
Beast Regards,
AR

evpopov commented:
/bot run formatting

kstribrnAmzn commented:
> I will try rebasing this PR to keep it up to date instead of merging

That's always my strategy so not a problem. Also thank you for the status update :)

evpopov commented:
Force-push log:
Originally, I wanted to keep as much of this PR's code as possible out of the original +TCP files, Therefore I put as much as I could in IGMP.c. There is also a function (prvSetMulticastSocketOption) that I wanted to keep private and not expose it to the user, so I placed it in IGMP.c and used an extern in the FreeRTOS_Sockets.c. 

This was not ideal, so I just moved the function in question to Sockets.c where it belongs and rebased to the latest main.

moninom1 commented:
Hi @evpopov 
Can you please help in moving the PR to drafts, as it is work in progress?

evpopov commented:
/bot run formatting

evpopov commented:
/bot run formatting

evpopov commented:
I was lucky enough to be able to work on this for almost the entire week. Check my latest commit description for details, but in summary, this PR now send IGMPv2 and MLD v1 Reports for all registered multicast groups.

One of my WiFi access points at work is doing some sort of multicast pruning and is probably running IGMP/MLD snooping but  It would never pass multicast messages to my device's solicited-node address. This resulted in my ethernet-connected device being inaccessible over IPv6 from PCs connected to the WiFi part of my office network. The PC's would try to send a neighbor solicitation to my device's solicited-node multicast address, but those messages would never reach the device. Once I got the device to send out MLD Reports, the WiFi AP now knew that my device was interested in receiving it's solicited-node multicast address and my IPv6 connectivity was miraculously fixed. 
The reason I'm mentioning this example is because I believe the IPv6 portion of FreeRTOS+TCP will simply not work in an environment that employs any kind of multicast pruning. I don't think it's just the solicited-node address that can get filtered out. I believe some of the name resolution protocols like LLMNR will not function either until the stack "declares" that it wants to receive FF02::1:3 by sending an MLD report for that group address. The described issue is not present in networks employing "dumb" switches and access-points.

Back to my update.... This is still very much a work in progress, but it is now sort of functional on both IPv4 and IPv6 and multiple interfaces and that is quite a step up from my original code-base. There is a lot of cleanup to be done and functionality to be added. For example, I need to add functionality for parsing MLD Queries that complements the existing IGMP query parser code. 

p.s. Originally I kept most changes in my added IGMP files, but as this PR gets more and more involved with IPv6 and multiple end-points, I'm starting to move functions and definitions from "my segregated IGMP files" to the regular code base. This makes the PR look more and more far-reaching because it affects a lot of files, but there's not much I can do about it. I'll do my best to squash multiple commits and keep this PR rebased on the latest main.

evpopov commented:
/bot run formatting

evpopov commented:
This PR now also works with IPv6 sockets.
The PR is still not ready for a proper evaluation and I more time for cleaning up and figuring out some issues with multiple interfaces. If anyone feels particularly adventurous though, here's some sample code.

**IPv4 quick and dirty example:**
```
uint32_t ulMCastAddress = 0xEFEFEF01;
struct freertos_sockaddr LocalAddr;
uint32_t LocalAddrLen = sizeof(LocalAddr);

//Setup the local end point
LocalAddr.sin_family = FREERTOS_AF_INET;
LocalAddr.sin_port = FreeRTOS_htons(63302);
LocalAddr.sin_address.ulIP_IPv4 = FreeRTOS_htonl(FREERTOS_INADDR_ANY);
LocalAddrLen

static Socket_t MCastRecvSock = FreeRTOS_socket( FREERTOS_AF_INET, FREERTOS_SOCK_DGRAM, FREERTOS_IPPROTO_UDP);
FreeRTOS_bind( MCastRecvSock, &LocalAddr, LocalAddrLen);

struct freertos_ip_mreq mreq;
mreq.imr_multiaddr.s_addr = FreeRTOS_htonl(ulMCastAddress);
mreq.imr_interface.s_addr = FreeRTOS_htonl(FREERTOS_INADDR_ANY);
FreeRTOS_setsockopt( MCastRecvSock, 0, FREERTOS_SO_IP_ADD_MEMBERSHIP, ( void * ) &mreq, sizeof( mreq ) );

```

**IPv6 quick and dirty example:**
```
const struct xIPv6_Address xV6MCastGroup = { { 0xFFU, 0x15U, 0x00U, 0x00U, 0x00U, 0x00U, 0x00U, 0x00U, 0x00U, 0x00U, 0x00U, 0x00U, 0x66U, 0x66U, 0x77U, 0x77U } };
struct freertos_ipv6_mreq mreq_v6;

struct freertos_sockaddr LocalAddr;
uint32_t LocalAddrLen = sizeof(LocalAddr);

//Setup the local end point
memset( &LocalAddr, 0x00, sizeof( struct freertos_sockaddr));
LocalAddr.sin_family = FREERTOS_AF_INET6;
LocalAddr.sin_port = FreeRTOS_htons(63306);
memcpy( LocalAddr.sin_address.xIP_IPv6.ucBytes, FreeRTOS_in6addr_any.ucBytes, ipSIZE_OF_IPv6_ADDRESS );

// Setup the struct used for adding/dropping multicast membership
memcpy( mreq_v6.ipv6mr_multiaddr.s6_addr.ucBytes, xV6MCastGroup.ucBytes, ipSIZE_OF_IPv6_ADDRESS );
mreq_v6.ipv6mr_interface = 0;

//Get a socket
static Socket_t MCastRecvSock = FreeRTOS_socket( FREERTOS_AF_INET6, FREERTOS_SOCK_DGRAM, FREERTOS_IPPROTO_UDP);
//Bind it
FreeRTOS_bind( MCastRecvSock, &LocalAddr, LocalAddrLen);
// Subscribe to the multicast group
FreeRTOS_setsockopt( MCastRecvSock, 0, FREERTOS_SO_IPV6_ADD_MEMBERSHIP, ( void * ) &mreq_v6, sizeof( mreq_v6 ) );
```

The code still needs work on how interfaces and sockets interact. The way I understand things with other operating systems and stacks is that in IPv4 we can subscribe to a multicast group on either a specific interface, or on all interfaces and the specific interface is designated by it's IP address.
In IPv6 however, we specify the interface by index and we can set multiple interfaces my calling setsockopt multiple times. Also index 0 is the "default" interface. 
This quickly becomes a mess so I think in this regard this PR may take a different approach, but I'm not sure what I'm going to do yet. I'm leaning towards allowing a socket ( either IPv4 or IPv6 ) to subscribe to a multicast group on either a specific interface or on all interfaces.
Another socket option that still needs implementing is IP_MULTICAST_IF  that specifies the outgoing interface when sending multicasts.

evpopov commented:
\bot run formatting

evpopov commented:
I updated the sample code above to reflect the latest changes. The changes are that the code now uses identical socket option names and option values for both IPv4 and IPv6. 
Subscribing and unsubscribing is achieved through the `IP_MReq_t` struct. It should be initialized with an IPv4 multicast group address when used with an AF_INET socket, or and IPv6 address when used with an AF_INET6 socket.
Setting the outgoing multicast TTL ( Hops Limit in the IPv6 terminology ) is done through a uint8_t pointer that holds the value of 0 to 255

The latest commit also adds a field to the NetworkBufferDescriptor_t. That field represents the outgoing TTL for packets. The field is added without #ifdef because it is used even with multicasts disabled. The latest commit populates the field in all possible situations except TCP, so that the proper TTL value is used based on the packet that is being sent. This new field is also used for the "Hop Limit" field which is the IPv6 equivalent of TTL

Still, don't look too closely into the details of the code. I have to rename the IGMP file to something more generic and I will also be squashing the commits soon.

evpopov commented:
/bot run formatting

tony-josi-aws commented on source/FreeRTOS_IGMP.c:
```diff
@@ -0,0 +1,1068 @@
+
+    if( pxNetworkBuffer->xDataLength < sizeof( IGMPPacket_t ) )
+    {
+        return eReturn;
```
Can we have single return statement per function?

evpopov replied:
@tony-josi-aws Yes, I'm aware of that and it is on my to-do list to go over things like returns, naming, casting, etc

amazonKamath replied:
I think we can help out on the MISRA violations once the code is stable.

evpopov commented:
Update:

- Sends IGMPv2 Leave Group messages whenever the last socket subscribed to a group drops that membership.
- Moves more functions to FreeRTOS_Sockets.c where they belong.
- Adds ipconfigPERIODIC_MULTICAST_REPORT_INTERVAL for debug purposes when there is no IGMP/MLD querier. Maybe we can remove that later.
- All commits squashed to keep all changes clearly visible.

I still have not implemented the NetworkInterface <-> MulticastGroup relationship. I will get back on this in 1 week.

evpopov commented:
Update:

Adds checks for valid network interface pointer when adding or dropping multicast membership.
Filters multicast messages from interfaces that the socket is not subscribed on.
Renames "IGMP" in all places where I should have used the more generic "Multicast"
Renames FreeRTOS_IGMP.c to FreeRTOS_IP_Multicast.c since it handles all multicasts reports.

evpopov commented:
/bot run formatting

evpopov commented:
/bot run formatting

evpopov commented:
I plan to add better IPv4 and IPv6 build separation.
I will continue keeping this PR in sync with #1065

evpopov commented:
Update:
As my last commit states, I just added some IPv4 / IPv6 build separation and rebased everything on the latest main to keep this PR clean. 
Honestly, this is very close to being ready for review but I'd rather wait until #1065 gets merged so that I can rebase on it. That will of course subtract #1065 's changes from here to make this PR a little cleaner.
The two big items on my todo list right now are:
- Sending the appropriate MLD message when a socket leaves a group
- Implementing a "default" outgoing interface for multicasts. I need to just sit down and think about it.... This may just boil down to using the first interface as default.

evpopov commented:
Update: This PR is now based on #1065.
Please do not merge main into this branch. I will take care of rebasing it as main and #1065 move along.

tbhavya443 commented:
Hi @evpopov I pulled the code from  MCast_PR,  for this we are working on the driver adsp-sc594 with the ip 100.64.10.5 for this ip we are able to get multicast messages from wireshark
when I added the multiple vlans like 100.64.10.5, 100.64.13.5 and 100.64.22.5 in the driver side we are able to get ping communication but we lost multicast messages as we enable ipconfigSUPPORT_IP_MULTICAST also I am not able to get multicast messages form wireshark 
Can you please help me out of this whether I need to enable anything in the code or it will support for multiple interfaces?

evpopov commented:
Hi @tbhavya443, 
I am not sure I understand your setup very well. FreeRTOS+TCP does not officially support VLANs, So for simplicity sake can you try and reproduce the issue without using VLANs. Please also do the following:

- What are your exact interface and endpoint settings. That will shed some light on the overall situation.
- Can you temporarily change your network driver to allow all packets in. In other words, put it in promiscuous mode and see if that helps.
- Make sure that every socket that needs multicasts, actually subscribes to the appropriate multicast group address. If you skip this step, then +TCP will not send multicast traffic to the sockets. There is an example for how to do that in here. Just scroll up or follow [this link to one of my previous comments](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/pull/1019#issuecomment-1779881686)

Hope that helps.

evpopov commented:
Since #1065 got merged recently, I have rebased this PR onto main and force-pushed so that all changes are clearly visible in a single commit.
This PR is almost ready for review, but sadly I will not have time to work on it in the upcoming month.
I will make sure to rebase it regularly so that it doesn't fall behind.

large-cat commented:
> Since #1065 got merged recently, I have rebased this PR onto main and force-pushed so that all changes are clearly visible in a single commit. This PR is almost ready for review, but sadly I will not have time to work on it in the upcoming month. I will make sure to rebase it regularly so that it doesn't fall behind.

Hi @evpopov I want use multicast in +tcp(ipv4), and I have rebase your MCast_PR branch onto main.
I’d like to know if there are any risks in doing this. Also, when will this feature be merged into the main branch?
Thank you, I look forward to your reply. This work is very useful to me.

amazonKamath commented:
> > Since #1065 got merged recently, I have rebased this PR onto main and force-pushed so that all changes are clearly visible in a single commit. This PR is almost ready for review, but sadly I will not have time to work on it in the upcoming month. I will make sure to rebase it regularly so that it doesn't fall behind.
> 
> Hi @evpopov I'd like to use multicast in +tcp(ipv4), and I have rebased your MCast_PR branch onto main. I’d like to know if there are any risks in doing this. Also, when will this feature be merged into the main branch? Thank you, I look forward to your reply. This work is very useful to me.

@large-cat The PR is not merged yet because  unit tests and memory safety proofs(CBMC) needs to be updated/added. We haven't had the chance to do it from our side yet due to other pressing priorities.  We would welcome any contribution on that side. Also, If you use this PR for testing the feature, please do share your feedback. Thank you.

evpopov commented:
@large-cat,
I haven't looked into the last 20-ish commits in great details, but there shouldn't be any dangers in rebasing this PR onto main. I will do this myself as soon as I get a chance. I have been very busy at work lately, and haven't been able to keep up with rebasing this as often as I'd like. I will do my best to carve out some time before the end of the year and bring this up to date with main.

With that being said, my production code uses this PR and I'm happy with how it works, so I think you are relatively safe using it.
As for when it will get merged, I just don't know. Off the top of my head, this PR is still missing the code to send IPv6 MLD reports which would prevent proper IPv6 multicast operation in networks that employ MLD snooping and multicast pruning, but in simple switched networks, IPv6 multicasts works fine.

As @amazonKamath mentioned, Id you end up using this PR or even if you just test it out, please give us your feedback.

large-cat commented:
> @large-cat, I haven't looked into the last 20-ish commits in great details, but there shouldn't be any dangers in rebasing this PR onto main. I will do this myself as soon as I get a chance. I have been very busy at work lately, and haven't been able to keep up with rebasing this as often as I'd like. I will do my best to carve out some time before the end of the year and bring this up to date with main.
> 
> With that being said, my production code uses this PR and I'm happy with how it works, so I think you are relatively safe using it. As for when it will get merged, I just don't know. Off the top of my head, this PR is still missing the code to send IPv6 MLD reports which would prevent proper IPv6 multicast operation in networks that employ MLD snooping and multicast pruning, but in simple switched networks, IPv6 multicasts works fine.
> 
> As @amazonKamath mentioned, Id you end up using this PR or even if you just test it out, please give us your feedback.
@evpopov 
Ok, Thank you.
I’m using this PR in my project. If I come across any issues during usage, I’ll be sure to report them back to you.

mpowellADC commented:
@evpopov Thank you guys for all of the work you've done adding Multicast, it's an important part of a project I'm working on.  I have to ask: is there any intention to support multiple groups on one socket?  Implementing an EtherNet/IP Adapter that can support multiple Scanners seems to demand that functionality, at least given certain local constraints.

evpopov commented:
@mpowellADC Thanks for the kind words! I'm very happy that people are finding this branch useful.
There is no plan to support multiple groups per socket. As a matter of fact, at the very beginning of this effort, every socket had a list of groups it was subscribed to. This however was causing the code to be a lot more complex. Things like adding groups, removing groups, and especially filtering incoming frames had to iterate over the subscribed groups list. After a discussion with the code maintainers, we decided that this was adding needless complexity for a very narrow use case and the idea was scrapped. 

It sounds like you may be developing an Ethernet/IP master and if that is the case, you could use unicast connections to minimize the load on the network. That will allow your server to use 1 socket to talk to as many slave devices as you have resources for.

mpowellADC commented:
@evpopov Thanks for the reply!  Unfortunately this project doesn't have the freedom to choose not to support something as standard as multicast listening on Ethernet/IP.  We may wind up having to hack that group list back in.

evpopov commented:
Removed the "unspecified source address" restriction from FreeRTOS_IPv6.c:500
 as discussed in #1124 and #1204

thirtytwobits commented on source/FreeRTOS_IP_Multicast.c:
```diff
@@ -0,0 +1,983 @@
+    uint32_t ulRandom;
+    MCastReportData_t * pxMRD;
+
+    xMaxCountdown = min( 1, xMaxCountdown );
```
is this supposed to be max? `xMaxCountdown` is an integer so min is meaningless here.


Also note that you've introduced `min` and `max` into the core source files of this project which are famously not portable. You probably don't want to do this.

evpopov replied:
Done. Also found and removed two more.

thirtytwobits commented on source/FreeRTOS_IP_Multicast.c:
```diff
@@ -0,0 +1,983 @@
+        {
+            /* IPv6 MLD */
+            pxIPv6_GroupAddress = ( const IPv6_Address_t * ) pvGroupAddress;
+            BaseType_t i, xSpecificQuery = pdFALSE;
```
`i` should be unsigned as `ipSIZE_OF_IPv6_ADDRESS` is unsigned.

evpopov replied:
Fixed by completely redoing the specific query check. Back when I wrote this I did not know FreeRTOS_in6addr_any existed.

thirtytwobits commented on source/FreeRTOS_IP_Multicast.c:
```diff
@@ -0,0 +1,983 @@
+/*
```
Need to add this file to the target sources for `freertos_plus_tcp` in `source/CMakeLists.txt`

evpopov replied:
Done

htibosch commented:
@evpopov, as this PR is in draft mode, it doesn't really bother if it stays long.
But looking again at the proposed changes, I think it is a good PR and it is worth to go through the checks and have it merged.

What do you think? Does it need more changes?

Integration testing will be a bit of work for me, as I don't have an application that needs multi-cast groups. That must be simulated.

htibosch commented:
@evpopov, if you want github to perform the checks, could you move it from "work in progress" to normal, and also get it up to date with main? Then we can see how much work is needed for "CI".
Thanks

shyne99 commented:
Sorry if this is not the place for this. Is there anything preventing this PR to be merged ? or at least be moved from draft to normal ?

evpopov commented:
Hi All,
I changed the status of this PR to ready-for-review.
I haven't been able to do any more development. As you may have noticed, every once in a while I have been re-basing this on main and force pushing and even though I haven't done any development on this for a while, I'm happy to report that I have been using this branch in production with my company's product with very good results.
Anyone willing to jump in and help with this PR is more than welcome. I do not have the time available to ensure all tests are covered.

This PR only supports a single a multicast group per socket in order to simplify the socket matching search on incoming packets. This could be a topic for further future improvements, but as I said earlier, I do not have the time to dedicate to this.

rawalexe commented:
Hello @evpopov 
Thank you so much for contributing to the repo and all the effort from 3 years. I will have a look sometime today and will try to get it merged.

htibosch commented:
Hi @evpopov , thanks for all the good work. I've looked with a critical eye at your code, and I have no complaints. Well except the testing code that you published, that's indeed quick and dirty :-)

I ran the tests on an STM32H7 and I have no complaints, even with 20 sockets that call `xEnlistMulticastReport()`/`vDelistMulticastReport()`, in a random order.

I used the standard function `pxUDPSocketLookup( )`:
~~~
{
    vTaskSuspendAll();
    xSocket = pxUDPSocketLookup( FreeRTOS_htons( usLocalPort ) );
    xTaskResumeAll();
}
~~~
to look up a UDP socket that is bound to `usLocalPort` (10001, 10002, 10003 etc).
With this, I don't have to keep track of all sockets that have been opened.

I use a .py script that gives this output:

~~~
igmp>python igmp_stagger.py
--- Starting IGMP Stagger Test on \Device\NPF_{A248D794-C977-4AF5-81DA-0E0563A6B2C4} ---
--- Max Response Time set to 10 seconds ---

Sending General Query to 224.0.0.1 (MRT: 10s)...
Waiting 12s for staggered responses...
  [REPORT] Group: 224.0.0.251 | Delay: 0.86s
  [REPORT] Group: 224.0.1.1 | Delay: 2.26s
  [REPORT] Group: 224.0.1.4 | Delay: 3.65s
  [REPORT] Group: 239.239.239.1 | Delay: 5.53s
  [REPORT] Group: 224.0.0.252 | Delay: 6.70s
  [REPORT] Group: 224.0.1.3 | Delay: 8.41s
  [REPORT] Group: 224.0.1.5 | Delay: 9.30s
  [REPORT] Group: 224.0.1.2 | Delay: 9.79s

Test complete.
~~~
Here is the Python script that I used:

[igmp_stagger.zip](https://github.com/user-attachments/files/31343181/igmp_stagger.zip)

Please let me know if I/we can assist you with something. Hein
